appinventor/components-ios/src/extensions/posenet.swift

```diff
@@ -0,0 +1,422 @@
+    _keyPoints["rightWrist"] = []
+    _keyPoints["leftHip"] = []
+    _keyPoints["rightHip"] = []
+    _keyPoints["leftKnee"] = []
+    _keyPoints["rightKnee"] = []
+    _keyPoints["leftAnkle"] = []
+    _keyPoints["rightAnkle"] = []
+    print("created posenet extension")
+  }
+  
+  /**
+     Configure the current webview. Add a WKScriptMessageHandler
+     to handle the following Javascript calls necessary for
+     Posenet Extension within userContentController(_ :didRecieve:):
+            ready()
+            reportImage(String dataURL)
+            reportResult(String result)
+            error(int errorCode, String errorMessage)
+     Equivalent to the @JavascriptInterface in the Android Posenet Extension.
+   */
+  private func configureWebView(_ webview: WKWebView) {
+    self._webview = webview
+    let config = webview.configuration
+    let controller = config.userContentController
+    controller.add(self, name: "PosenetExtension")
+  }
+  
+  
+  /**
+    Implements WKUIDelegate protocol to handle camera and video permissions.
+  */
+  @available(iOS 15.0, *)
+  public func webView(_ webView: WKWebView,
+              requestMediaCapturePermissionFor
+              origin: WKSecurityOrigin,initiatedByFrame
+              frame: WKFrameInfo,type: WKMediaCaptureType,
+              decisionHandler: @escaping (WKPermissionDecision) -> Void) {
+                decisionHandler(.grant)
+  }
+  
+  /**
+   Configures Javascript interface. TFJS app.js file has been
+   edited such that it directly calls userContentController.
+   Generic implementation yet to be built.
+   */
+  open func userContentController(
+    _ userContentController: WKUserContentController,
+    didReceive message: WKScriptMessage
+  ) {
+    print("recieving content")
+    if message.name == "PosenetExtension" {
+      guard let dict = message.body as? [String: Any],
+            let functionCall = dict["functionCall"] as? String,
+            let args = dict["args"] else {
+        print("JSON Error message not recieved")
+        return
+      }
+      if functionCall == "ready" {
+        print("Model Ready")
+        ModelReady()
+        if _enabled {
+          MinPartConfidence = _minPartConfidence
+          MinPoseConfidence = _minPoseConfidence
+          UseCamera = _cameraMode
+        }
+      }
+      if functionCall == "reportResult" {
+        print("Reporting Result")
+        do {
+          let result = try getYailObjectFromJson(args as? String, true)
+          print(result)
+        } catch {
+          print("Error parsing JSON from web view")
+          Error(_ERROR_JSON_PARSE_FAILED as AnyObject, "Error parsing JSON from web view" as AnyObject)
+        }
+      }
+      if functionCall == "reportImage" {
+        print("Reporting Image")
+        // baackground image?
+        VideoUpdated()
+      }
+      if functionCall == "error" {
+        // TODO: Error function
+        //        let (errorCode, errorMessage) = args
```

Comment:
I don't think you could do a destructuring bind in this way in swift from JSON since the types are not known at compile time. However, you could add a guard let with multiple definitions:

```swift
guard let argsList = args as? NSArray,
    let errorCode = argsList[0] as? NSNumber,
    let errorMessage = argsList[1] as? NSString else {
  ...
}
```

Reply:
Apple has a good article on JSON and Swift here: https://developer.apple.com/swift/blog/?id=37